### ISSUE TITLE
Custom API routes and controllers return 404 not found

### ISSUE THREAD

**Issue by sudopower** [opensupports/opensupports#525]
Hey, It would be really nice if someone could help me out with this,
I'm new to React. I'm getting a 404 on API Call for newly created Route. 

I've modified the Department Model, Controller and admin-panel-department.js to add a couple of fields. They are working fine.

Now inside of staff section just like department I added another section.

![issue](https://user-images.githubusercontent.com/14028003/56206592-604d7680-606a-11e9-883b-94ce0d5aa89a.png)
 
I am a beginner please bear with me and let me know what additional information can I include. I've just added a couple of lines here and there and commented them so that you don't have to go through the entire file.

> ROUTES               Routes.js

```
import React from 'react';
import {Router, Route, IndexRoute, IndexRedirect} from 'react-router';

import App from 'app/App';
import DemoPage from 'app/demo/components-demo-page';
import history from 'lib-app/history';

import MainLayout from 'app/main/main-layout';
import MainHomePage from 'app/main/main-home/main-home-page';
import MainSignUpPage from 'app/main/main-signup/main-signup-page';
import MainVerifyTokenPage from 'app/main/main-verify-token-page';
import MainRecoverPasswordPage from 'app/main/main-recover-password/main-recover-password-page';
import MainMaintenancePage from 'app/main/main-maintenance-page';
import MainCheckTicketPage from 'app/main/main-check-ticket-page';
import MainViewTicketPage from 'app/main/main-view-ticket-page';

import DashboardLayout from 'app/main/dashboard/dashboard-layout';
import DashboardListTicketsPage from 'app/main/dashboard/dashboard-list-tickets/dashboard-list-tickets-page';
import DashboardListArticlesPage from 'app/main/dashboard/dashboard-list-articles/dashboard-list-articles-page';
import DashboardCreateTicketPage from 'app/main/dashboard/dashboard-create-ticket/dashboard-create-ticket-page';
import DashboardEditProfilePage from 'app/main/dashboard/dashboard-edit-profile/dashboard-edit-profile-page';
import DashboardArticlePage from 'app/main/dashboard/dashboard-article/dashboard-article-page';
import DashboardTicketPage from 'app/main/dashboard/dashboard-ticket/dashboard-ticket-page';

// ADMIN PANEL
import AdminLoginPage from 'app/admin/admin-login-page';
import AdminPanelLayout from 'app/admin/panel/admin-panel-layout';

import AdminPanelStats from 'app/admin/panel/dashboard/admin-panel-stats';
import AdminPanelActivity from 'app/admin/panel/dashboard/admin-panel-activity';
import AdminPanelMyAccount from 'app/admin/panel/dashboard/admin-panel-my-account';

import AdminPanelMyTickets from 'app/admin/panel/tickets/admin-panel-my-tickets';
import AdminPanelNewTickets from 'app/admin/panel/tickets/admin-panel-new-tickets';
import AdminPanelAllTickets from 'app/admin/panel/tickets/admin-panel-all-tickets';
import AdminPanelViewTicket from 'app/admin/panel/tickets/admin-panel-view-ticket';
import AdminPanelCustomResponses from 'app/admin/panel/tickets/admin-panel-custom-responses';
import AdminPanelCustomTags from 'app/admin/panel/tickets/admin-panel-custom-tags';

import AdminPanelListUsers from 'app/admin/panel/users/admin-panel-list-users';
import AdminPanelViewUser from 'app/admin/panel/users/admin-panel-view-user';
import AdminPanelBanUsers from 'app/admin/panel/users/admin-panel-ban-users';
import AdminPanelCustomFields from 'app/admin/panel/users/admin-panel-custom-fields';

import AdminPanelListArticles from 'app/admin/panel/articles/admin-panel-list-articles';
import AdminPanelViewArticle from 'app/admin/panel/articles/admin-panel-view-article';

import AdminPanelStaffMembers from 'app/admin/panel/staff/admin-panel-staff-members';
import AdminPanelDepartments from 'app/admin/panel/staff/admin-panel-departments';
//**************************CHANGES***********************************
import AdminPanelSectors from 'app/admin/panel/staff/admin-panel-sectors';
//**************************CHANGES***********************************
import AdminPanelViewStaff from 'app/admin/panel/staff/admin-panel-view-staff';

import AdminPanelSystemPreferences from 'app/admin/panel/settings/admin-panel-system-preferences';
import AdminPanelAdvancedSettings from 'app/admin/panel/settings/admin-panel-advanced-settings';
import AdminPanelEmailSettings from 'app/admin/panel/settings/admin-panel-email-settings';

// INSTALLATION
import InstallLayout from 'app/install/install-layout';
import InstallStep1Language from 'app/install/install-step-1-language';
import InstallStep2Requirements from 'app/install/install-step-2-requirements';
import InstallStep3Database from 'app/install/install-step-3-database';
import InstallStep4UserSystem from 'app/install/install-step-4-user-system';
import InstallStep5Settings from 'app/install/install-step-5-settings';
import InstallStep6Admin from 'app/install/install-step-6-admin';
import InstallStep7Completed from 'app/install/install-step-7-completed';

export default (
    <Router history={history}>
        <Route component={App}>
            <Route path='/' component={MainLayout}>
                <IndexRoute component={MainHomePage} />
                <Route path='signup' component={MainSignUpPage}/>
                <Route path='verify-token/:email/:token' component={MainVerifyTokenPage}/>
                <Route path='recover-password' component={MainRecoverPasswordPage}/>
                <Route path='maintenance' component={MainMaintenancePage}/>

                <Route path='create-ticket' component={DashboardCreateTicketPage}/>
                <Route path='check-ticket(/:ticketNumber/:email)' component={MainCheckTicketPage}/>
                <Route path='view-ticket/:ticketNumber' component={MainViewTicketPage}/>
                <Route path='articles' component={DashboardListArticlesPage}/>
                <Route path='article/:articleId' component={DashboardArticlePage}/>

                <Route path='dashboard' component={DashboardLayout}>
                    <IndexRoute component={DashboardListTicketsPage} />
                    <Route path='articles' component={DashboardListArticlesPage}/>

                    <Route path='create-ticket' component={DashboardCreateTicketPage}/>
                    <Route path='edit-profile' component={DashboardEditProfilePage}/>

                    <Route path='article/:articleId' component={DashboardArticlePage}/>
                    <Route path='ticket/:ticketNumber' component={DashboardTicketPage}/>
                </Route>
            </Route>
            <Route path="install" component={InstallLayout}>
                <IndexRedirect to="step-1" />
                <Route path="step-1" component={InstallStep1Language}/>
                <Route path="step-2" component={InstallStep2Requirements} />
                <Route path="step-3" component={InstallStep3Database} />
                <Route path="step-4" component={InstallStep4UserSystem} />
                <Route path="step-5" component={InstallStep5Settings} />
                <Route path="step-6" component={InstallStep6Admin} />
                <Route path="step-7" component={InstallStep7Completed} />
            </Route>
            <Route path="admin">
                <IndexRoute component={AdminLoginPage} />
                <Route path="panel" component={AdminPanelLayout}>
                    <IndexRedirect to="activity" />
                    <Route path="stats" component={AdminPanelStats} />
                    <Route path="activity" component={AdminPanelActivity} />
                    <Route path="my-account" component={AdminPanelMyAccount} />

                    <Route path="tickets">
                        <IndexRedirect to="my-tickets" />
                        <Route path="my-tickets" component={AdminPanelMyTickets} />
                        <Route path="new-tickets" component={AdminPanelNewTickets} />
                        <Route path="all-tickets" component={AdminPanelAllTickets} />
                        <Route path="custom-responses" component={AdminPanelCustomResponses} />
                        <Route path="view-ticket/:ticketNumber" component={AdminPanelViewTicket} />
                        <Route path="custom-tags" component={AdminPanelCustomTags} />
                    </Route>

                    <Route path="users">
                        <IndexRedirect to="list-users" />
                        <Route path="list-users" component={AdminPanelListUsers} />
                        <Route path="view-user/:userId" component={AdminPanelViewUser} />
                        <Route path="ban-users" component={AdminPanelBanUsers} />
                        <Route path="custom-fields" component={AdminPanelCustomFields} />
                    </Route>

                    <Route path="articles">
                        <IndexRedirect to="list-articles" />
                        <Route path="list-articles" component={AdminPanelListArticles} />
                        <Route path="view-article/:articleId" component={AdminPanelViewArticle} />
                    </Route>

                    <Route path="staff">
                        <IndexRedirect to="staff-members" />
                        <Route path="staff-members" component={AdminPanelStaffMembers} />
                        <Route path="view-staff/:staffId" component={AdminPanelViewStaff} />
                        <Route path="departments" component={AdminPanelDepartments} />
                        //**************************CHANGES***********************************
                        <Route path="sectors" component={AdminPanelSectors} />
                        //**************************CHANGES***********************************
                    </Route>

                    <Route path="settings">
                        <IndexRedirect to="system-preferences" />
                        <Route path="system-preferences" component={AdminPanelSystemPreferences} />
                        <Route path="advanced-settings" component={AdminPanelAdvancedSettings} />
                        <Route path="email-settings" component={AdminPanelEmailSettings} />
                    </Route>
                </Route>
            </Route>

            <Route name='Demo' path='components-demo' component={DemoPage} />
        </Route>
    </Router>
);


```

> CONTROLLER               /server/controllers/system/add-sector.php
```<?php
use Respect\Validation\Validator as DataValidator;

/**
 * @api {post} /system/add-sector Add sector
 * @apiVersion 4.4.0
 *
 * @apiName Add sector
 *
 * @apiGroup System
 *
 * @apiDescription This path create a new sector.
 *
 * @apiPermission staff3
 *
 * @apiParam {String} name Name of the new sector.
 *
 * @apiUse NO_PERMISSION
 *
 * @apiSuccess {Object} data Empty object
 *
 */

class AddSectorController extends Controller {
    const PATH = '/add-sector';
    const METHOD = 'POST';
    public function validations() {
        return [
            'permission' => 'staff_3',
            'requestData' => [
                'name' => [
                    'validation' => DataValidator::length(2, 100),
                    'error' => ERRORS::INVALID_NAME
                ]
            ]
        ];
    }

    public function handler() {
        $name = Controller::request('name');

        $sectorInstance = new Sector();

        $sectorInstance->setProperties([
            'name' => $name
        ]);
        $sectorInstance->store();

        Log::createLog('ADD_SECTOR', $name);

        Response::respondSuccess();

    }
}
```

> MODEL               server/models/Sector.php
```
<?php
use RedBeanPHP\Facade as RedBean;

/**
 * @api {OBJECT} Department Sector
 * @apiVersion 4.4.0
 * @apiGroup Data Structures
 * @apiParam {Number} id Id of the sector.
 * @apiParam {String} name Name of the sector.
 * @apiParam {[Staff](#api-Data_Structures-ObjectStaff)[]} owners List of owners of the sector.
 * @apiParam {Boolean} private Indicates if the sector is not shown to users.
 */
 
class Sector extends DataStore {
    const TABLE = 'sector';

    public static function getProps() {
        return [
            'name',
            'sharedTicketList',
            'owners'
        ];
    }

    public function getDefaultProps() {
        return [
            'owners' => 0
        ];
    }

    public static function getAllSectorNames() {
        $sectorList = RedBean::findAll(Sector::TABLE);
        $sectorsNameList = [];

        foreach($sectorsList as $sector) {
            $sectorsNameList[] = [
                'id' => $sector->id,
                'name' => $sector->name
            ];
        }

        return $sectorsNameList;
    }

    public static function getPublicSectorNames() {
        $sectorsList = RedBean::findAll(Sector::TABLE);
        $sectorsNameList = [];

        foreach($sectorsList as $sector) {
            if(!$sector->private) {
                $sectorsNameList[] = [
                    'id' => $sector->id,
                    'name' => $sector->name
                ];
            }
        }

        return $sectorsNameList;
    }

    public function toArray() {
        return [
            'id' => $this->id,
            'name' => $this->name
        ];
    }
}
```

> The Error that I get in chrome console
```
xhr.js:178 POST http://localhost:3000/api/system/add-sector 404 (Not Found)
```
> CLIENT PAGE /client/src/admin/panel/staff/admin-panel-sector.js
```
import React from 'react';
import _ from 'lodash';
import {connect}  from 'react-redux';

import i18n from 'lib-app/i18n';
import API from 'lib-app/api-call';
import ConfigActions from 'actions/config-actions';

import AreYouSure from 'app-components/are-you-sure';
import DepartmentDropDown from 'app-components/department-dropdown';

import InfoTooltip from 'core-components/info-tooltip';
import Button from 'core-components/button';
import Header from 'core-components/header';
import Listing from 'core-components/listing';
import Form from 'core-components/form';
import FormField from 'core-components/form-field';
import SubmitButton from 'core-components/submit-button';
import DropDown from 'core-components/drop-down';
import Icon from 'core-components/icon';
import Message from 'core-components/message';

class AdminPanelSectors extends React.Component {
    static defaultProps = {
        items: []
    };

    state = {
        formLoading: false,
        selectedIndex: -1,
        selectedDropDownIndex: 0,
        edited: false,
        errorMessage: null,
        errors: {},
        form: {
            title: '',
            language: 'en',
            private: 0,
        }
    };

    render() {
        return (
            <div className="admin-panel-departments">
                <Header title={i18n('SECTOR')} description={i18n('SECTOR_DESCRIPTION')} />
                <div className="row">
                    <div className="col-md-4">
                        <Listing {...this.getListingProps()}/>
                    </div>
                    <div className="col-md-8">
                        {(this.state.errorMessage) ? <Message type="error">{i18n(this.state.errorMessage)}</Message> : null}
                        <Form {...this.getFormProps()}>
                            <div>
                                <FormField className="admin-panel-sectors__name" label={i18n('NAME')} name="name" validation="NAME" required fieldProps={{size: 'large'}}/>
                            </div>

                            <SubmitButton size="medium" className="admin-panel-departments__update-name-button" type="secondary">
                                {i18n((this.state.selectedIndex !== -1) ? 'UPDATE_SECTOR' : 'ADD_SECTOR')}
                            </SubmitButton>
                        </Form>
                        {(this.state.selectedIndex !== -1 && this.props.departments.length) ? this.renderOptionalButtons() : null}
                    </div>
                </div>
            </div>
        );
    }

    renderOptionalButtons() {
        return (
            <div className="admin-panel-departments__optional-buttons">
                <div className="admin-panel-departments__discard-button">
                    <Button onClick={this.onDiscardChangesClick.bind(this)} size="medium">{i18n('DISCARD_CHANGES')}</Button>
                </div>
                {this.props.departments.length > 1 ? this.renderDeleteButton() : null}
            </div>
        );
    }

    renderDeleteButton() {
        return (
            <div className="admin-panel-departments__delete-button">
                <Button onClick={this.onDeleteClick.bind(this)} size="medium">{i18n('DELETE')}</Button>
            </div>
        );
    }

    renderDelete() {
        return (
            <div>
                {i18n('WILL_DELETE_DEPARTMENT')}
                <div className="admin-panel-departments__transfer-tickets">
                    <span className="admin-panel-departments__transfer-tickets-title">{i18n('TRANSFER_TICKETS_TO')}</span>
                    <DepartmentDropDown className="admin-panel-departments__transfer-tickets-drop-down" departments={this.getDropDownDepartments()} onChange={(event) => this.setState({selectedDropDownIndex: event.index})} size="medium"/>
                </div>
            </div>
        );
    }

    getListingProps() {
        return {
            className: 'admin-panel-departments__list',
            title: i18n('SECTOR'),
            items: this.props.departments.map(department => {
                return {
                    content: (
                        <span>
                            {department.name}
                            {department.private*1 ? <Icon className="admin-panel-departments__private-icon" name='user-secret'/> : null }
                            {(!department.owners) ? (
                                <span className="admin-panel-departments__warning">
                                    <InfoTooltip type="warning" text={i18n('NO_STAFF_ASSIGNED')}/>
                                </span>
                            ) : null}
                        </span>
                    )
                };
            }),
            selectedIndex: this.state.selectedIndex,
            enableAddNew: true,
            onChange: this.onItemChange.bind(this),
            onAddClick: this.onItemChange.bind(this, -1)
        };
    }

    getFormProps() {
        return {
            values: this.state.form,
            errors: this.state.errors,
            loading: this.state.formLoading,
            onChange: (form) => {this.setState({form, edited: true})},
            onValidateErrors: (errors) => {this.setState({errors})},
            onSubmit: this.onFormSubmit.bind(this)
        };
    }

    onItemChange(index) {
        if(this.state.edited) {
            AreYouSure.openModal(i18n('WILL_LOSE_CHANGES'), this.updateForm.bind(this, index));
        } else {
            this.updateForm(index);
        }
    }

    onFormSubmit(form) {
        this.setState({formLoading: true, edited: false});


        if(this.state.selectedIndex !== -1) {
            API.call({
                path: '/system/edit-department',
                data: {
                    departmentId: this.getCurrentDepartment().id,
                    name: form.name,
                    custom1: form.custom1,
                    private: form.private ? 1 : 0
                }
            }).then(() => {
                this.setState({formLoading: false});
                this.retrieveDepartments();
            }).catch(result => this.setState({formLoading: false, errorMessage: result.message}));
        } else {
            API.call({
                path: '/system/add-sector',
                data: {
                    name: form.name
                }
            }).then(() => {
                this.retrieveDepartments();
                this.onItemChange(-1);
            }).catch(this.onItemChange.bind(this, -1));
        }
    }

    onDiscardChangesClick() {
        this.onItemChange(this.state.selectedIndex);
    }

    onDeleteClick() {
        this.setState({
            selectedDropDownIndex: 0
        });

        AreYouSure.openModal(this.renderDelete(), this.deleteDepartment.bind(this));
    }

    deleteDepartment() {
        API.call({
            path: '/system/delete-department',
            data: {
                departmentId: this.getCurrentDepartment().id,
                transferDepartmentId: this.getDropDownItemId()
            }
        }).then(() => {
            this.retrieveDepartments();
            this.onItemChange(-1);
        })
        .catch(result => this.setState({errorMessage: result.message}));
    }

    updateForm(index) {
        let form = _.clone(this.state.form);
        let department = this.getCurrentDepartment(index);

        form.name = (department && department.name) || '';
        form.custom1 = (department && department.custom1) || '';
        form.private = (department && department.private) || 0;

        this.setState({
            selectedIndex: index,
            edited: false,
            formLoading: false,
            form,
            errorMessage: null,
            errors: {}
        });
    }

    retrieveDepartments() {
        this.props.dispatch(ConfigActions.updateData());
        this.setState({
            edited: false
        });
    }

    getCurrentDepartment(index) {
        return this.props.departments[(index == undefined) ? this.state.selectedIndex : index];
    }

    getDropDownItemId() {
        return this.props.departments.filter((department, index) => index !== this.state.selectedIndex)[this.state.selectedDropDownIndex].id;
    }

    getDropDownDepartments() {
        return this.props.departments.filter((department, index) => index !== this.state.selectedIndex);
    }
}

export default connect((store) => {
    return {
        departments: store.config.departments
    };
})(AdminPanelSectors);
```

# TL;DR
1. Duplicated Department Controller, Model, admin-panel-department.js (files included above)
2. Replaced all names and got the client working
3. Now API call returns 404 on console

**Comment by sudopower** [comment]
Okay this is embarrasing
I did not route to Controller inside of api router controller/system.php
Added ```$systemControllerGroup->addController(new AddSectorController);```
